File: frontend/src/stores/knowledge.ts
/**
 * 知識庫與文件的狀態（1E-2；03 §2、09 §2.3）。
 *
 * views 不直接 fetch（03 §1），所以 KB 清單與「當前 KB 的文件清單」這兩份資料
 * 只有這裡持有。它對三件事負責：
 *
 * 1. **本地清單與後端一致**：建立／更名／刪除／重跑之後就地更新，不重抓整份清單。
 *    重抓不只是多一次往返——那段空窗期畫面會閃回舊資料，使用者會以為沒生效而再按一次。
 * 2. **切換 KB 時不錯置**：慢的回應不得覆蓋後來的選擇（見 `fetchDocuments` 的代號守門）。
 * 3. **輪詢的依據**：`hasPendingDocuments` 是「還要不要問」的唯一來源。
 *
 * **錯誤一律往上拋**（`ApiError`），由 view 決定顯示方式。store 不吞錯，也不自己
 * 存一份 errorMessage——那會讓兩個頁面共用同一個過期的錯誤字串。
 */
import { defineStore } from 'pinia'
import { computed, ref } from 'vue'

import { request } from '@/api/client'
import { uploadDocument as uploadToApi } from '@/services/uploadService'
import type {
  DocumentListOut,
  DocumentOut,
  KnowledgeBaseCreateIn,
  KnowledgeBaseListOut,
  KnowledgeBaseOut,
  KnowledgeBaseUpdateIn,
} from '@/types/models'
import { isDocumentSettled } from '@/utils/documentStatus'

/**
 * ETL 進度的輪詢間隔。
 *
 * 3 秒是「感覺得到在動」與「不要打擾後端」之間的起手值（13 §1.2：文件值是起始點）。
 * 它是**可調參數**，之後要搬進統一設定畫面（15 §4.1、13 §3 帶進 1E 的缺口③）——
 * 先收成單一具名匯出，那時只要改這裡的來源，不必在各個 view 裡找散落的 3000。
 */
export const DOCUMENT_POLL_INTERVAL_MS = 3_000

const JSON_HEADERS = { 'Content-Type': 'application/json' } as const

export const useKnowledgeStore = defineStore('knowledge', () => {
  const knowledgeBases = ref<KnowledgeBaseOut[]>([])
  const documents = ref<DocumentOut[]>([])
  const currentKbId = ref<string | null>(null)
  const loadingBases = ref(false)
  const loadingDocuments = ref(false)

  /**
   * 文件清單的請求代號。切 KB 是一秒內會做兩次的動作，而回應不保證照順序回來
   * ——沒有這個守門的話，A 的慢回應會蓋掉 B 的清單：標題是 B、內容是 A，完全不報錯。
   */
  let documentsRequestId = 0

  const hasPendingDocuments = computed(() =>
    documents.value.some((document) => !isDocumentSettled(document.status)),
  )

  // ── 知識庫 ──────────────────────────────────────────────────────────────

  async function fetchKnowledgeBases(): Promise<void> {
    loadingBases.value = true
    try {
      const page = await request<KnowledgeBaseListOut>('/api/v1/knowledge-bases')
      knowledgeBases.value = page.items
    } finally {
      // finally 而不是成功路徑：少了它，失敗後的畫面會永遠轉圈，
      // 而重試按鈕通常就藏在那個轉圈的下面。
      loadingBases.value = false
    }
  }

  async function createKnowledgeBase(input: KnowledgeBaseCreateIn): Promise<KnowledgeBaseOut> {
    const created = await request<KnowledgeBaseOut>('/api/v1/knowledge-bases', {
      method: 'POST',
      headers: JSON_HEADERS,
      body: JSON.stringify(input),
    })
    // 排最前面：後端的清單是 `-created_at`（repositories/knowledge.py 的
    // `list_for_tenant`），加在最後的話重新整理之後它會自己跳到第一列。
    knowledgeBases.value = [created, ...knowledgeBases.value]
    return created
  }

  async function updateKnowledgeBase(
    kbId: string,
    patch: KnowledgeBaseUpdateIn,
  ): Promise<KnowledgeBaseOut> {
    const updated = await request<KnowledgeBaseOut>(
      `/api/v1/knowledge-bases/${encodeURIComponent(kbId)}`,
      { method: 'PATCH', headers: JSON_HEADERS, body: JSON.stringify(patch) },
    )
    // 就地替換：更名後跳到清單最後，會讓人以為自己改到了別的東西。
    knowledgeBases.value = knowledgeBases.value.map((kb) => (kb.id === kbId ? updated : kb))
    return updated
  }

  async function deleteKnowledgeBase(kbId: string): Promise<void> {
    await request<null>(`/api/v1/knowledge-bases/${encodeURIComponent(kbId)}`, { method: 'DELETE' })
    knowledgeBases.value = knowledgeBases.value.filter((kb) => kb.id !== kbId)
    if (currentKbId.value === kbId) {
      // 刪掉的正是現在開著的那個：留著文件清單會讓輪詢繼續問一個不存在的 KB。
      currentKbId.value = null
      documents.value = []
    }
  }

  // ── 文件 ────────────────────────────────────────────────────────────────

  /** `silent` 給輪詢用：翻動 loading 旗標的話，表格每 3 秒閃一次骨架畫面。 */
  async function fetchDocuments(kbId: string, options: { silent?: boolean } = {}): Promise<void> {
    const silent = options.silent === true
    documentsRequestId += 1
    const requestId = documentsRequestId
    currentKbId.value = kbId
    if (!silent) {
      loadingDocuments.value = true
    }
    try {
      const page = await request<DocumentListOut>(
        `/api/v1/knowledge-bases/${encodeURIComponent(kbId)}/documents`,
      )
      if (requestId !== documentsRequestId) {
        return // 使用者已經換到別的 KB，這份回應過期了
      }
      documents.value = page.items
    } finally {
      if (requestId === documentsRequestId && !silent) {
        loadingDocuments.value = false
      }
    }
  }

  /**
   * 輪詢用的重抓。沒有選定 KB 就什麼都不做——輪詢的 callback 可能在導航之後才跑到，
   * 那時對 `/knowledge-bases/undefined/documents` 發請求會拿到 404，畫面上會冒出
   * 一個與使用者動作無關的錯誤。
   */
  async function refreshDocuments(): Promise<void> {
    const kbId = currentKbId.value
    if (kbId === null) {
      return
    }
    await fetchDocuments(kbId, { silent: true })
  }

  async function uploadDocument(kbId: string, file: File): Promise<DocumentOut> {
    const created = await uploadToApi(kbId, file)
    if (kbId === currentKbId.value) {
      // 排最前面：剛傳完的那份是使用者現在唯一在看的東西。
      documents.value = [created, ...documents.value]
    }
    return created
  }

  async function deleteDocument(documentId: string): Promise<void> {
    // 先打 API 再移除（不做樂觀刪除）：403/409 之後畫面少一列、重新整理又出現，
    // 那比多等一個往返難解釋得多。
    await request<null>(`/api/v1/documents/${encodeURIComponent(documentId)}`, { method: 'DELETE' })
    documents.value = documents.value.filter((document) => document.id !== documentId)
  }

  /** 重新處理：後端回 202 + 已回到 `uploaded`、`doc_version` +1 的文件。 */
  async function reingestDocument(documentId: string): Promise<DocumentOut> {
    const updated = await request<DocumentOut>(
      `/api/v1/documents/${encodeURIComponent(documentId)}/reingest`,
      { method: 'POST' },
    )
    documents.value = documents.value.map((document) =>
      document.id === documentId ? updated : document,
    )
    return updated
  }

  return {
    knowledgeBases,
    documents,
    currentKbId,
    loadingBases,
    loadingDocuments,
    hasPendingDocuments,
    fetchKnowledgeBases,
    createKnowledgeBase,
    updateKnowledgeBase,
    deleteKnowledgeBase,
    fetchDocuments,
    refreshDocuments,
    uploadDocument,
    deleteDocument,
    reingestDocument,
  }
})

File: backend/api/v1/knowledge.py
"""`/knowledge-bases` 與 `/documents` 端點（09 §2.3）。

鐵則 3（Controller 三行原則）：解析請求 → 呼叫一個 Service 方法 → 回傳。

**權限宣告寫在每個端點上**（10 §3）：讀這個檔案就知道每條路需要什麼權限。分級的
依據是**破壞範圍**，不是動作類型：

- ``knowledge:read``——看。全部四個角色都有；Viewer 看不到文件清單的話，連「這個
  答案引用的是哪份文件」都無從確認。
- ``knowledge:write``——文件進出。Editor 的日常工作，破壞範圍限單一文件且是軟刪除。
- ``knowledge:admin``——建立/刪除 KB、改設定。刪 KB 連帶整庫文件失效，改 config 會
  讓整個知識庫需要重算——維運等級的動作。

兩個 router 合在一份檔案：它們是同一個 bounded context 的對外面，而 `/documents`
的路徑刻意**不掛在 KB 底下**（文件 id 已經唯一，巢狀路徑會逼 client 記住它屬於哪個
KB，而那在文件搬移時就錯了）。
"""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, File, UploadFile, status

from api.dependencies.auth import Principal
from api.dependencies.permissions import RequireScope
from api.schemas.knowledge import (
    DocumentListOut,
    DocumentOut,
    KnowledgeBaseCreateIn,
    KnowledgeBaseListOut,
    KnowledgeBaseOut,
    KnowledgeBaseUpdateIn,
)
from api.schemas.problem import ERROR_RESPONSES
from core.db import run_orm
from services.knowledge.documents import DocumentService, DocumentView
from services.knowledge.knowledge_bases import KnowledgeBaseService, KnowledgeBaseView

router = APIRouter(tags=["knowledge"], responses=ERROR_RESPONSES)
_knowledge_bases = KnowledgeBaseService()
_documents = DocumentService()


def _kb_out(view: KnowledgeBaseView) -> KnowledgeBaseOut:
    return KnowledgeBaseOut(
        id=view.id,
        name=view.name,
        description=view.description,
        status=view.status,
        document_count=view.document_count,
    )


def _document_out(view: DocumentView) -> DocumentOut:
    return DocumentOut(
        id=view.id,
        kb_id=view.kb_id,
        filename=view.filename,
        mime_type=view.mime_type,
        size_bytes=view.size_bytes,
        status=view.status,
        doc_version=view.doc_version,
        error=view.error,
    )


# ── 知識庫 ──────────────────────────────────────────────────────


@router.get("/knowledge-bases", operation_id="knowledge_bases_list")
async def list_knowledge_bases(
    principal: Annotated[Principal, Depends(RequireScope("knowledge:read"))],
) -> KnowledgeBaseListOut:
    views = await run_orm(_knowledge_bases.list_knowledge_bases, principal.tenant_id)
    return KnowledgeBaseListOut(items=[_kb_out(view) for view in views])


@router.post(
    "/knowledge-bases",
    operation_id="knowledge_bases_create",
    status_code=status.HTTP_201_CREATED,
)
async def create_knowledge_base(
    payload: KnowledgeBaseCreateIn,
    principal: Annotated[Principal, Depends(RequireScope("knowledge:write"))],
) -> KnowledgeBaseOut:
    """建立 KB 是 ``knowledge:write``，刪除與改設定才是 ``admin``。

    不對稱是刻意的：多一個空知識庫沒有破壞性，而 Editor 建不了 KB 的話，每次要新
    主題都得找管理員——那會讓人把所有東西倒進同一個 KB，檢索品質跟著爛掉。
    """
    view = await run_orm(
        _knowledge_bases.create,
        principal.tenant_id,
        name=payload.name,
        description=payload.description,
    )
    return _kb_out(view)


@router.get("/knowledge-bases/{kb_id}", operation_id="knowledge_bases_get")
async def get_knowledge_base(
    kb_id: uuid.UUID,
    principal: Annotated[Principal, Depends(RequireScope("knowledge:read"))],
) -> KnowledgeBaseOut:
    return _kb_out(await run_orm(_knowledge_bases.get, principal.tenant_id, kb_id))


@router.patch("/knowledge-bases/{kb_id}", operation_id="knowledge_bases_update")
async def update_knowledge_base(
    kb_id: uuid.UUID,
    payload: KnowledgeBaseUpdateIn,
    principal: Annotated[Principal, Depends(RequireScope("knowledge:admin"))],
) -> KnowledgeBaseOut:
    view = await run_orm(
        _knowledge_bases.update,
        principal.tenant_id,
        kb_id,
        name=payload.name,
        description=payload.description,
    )
    return _kb_out(view)


@router.delete(
    "/knowledge-bases/{kb_id}",
    operation_id="knowledge_bases_delete",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_knowledge_base(
    kb_id: uuid.UUID,
    principal: Annotated[Principal, Depends(RequireScope("knowledge:admin"))],
) -> None:
    await run_orm(_knowledge_bases.delete, principal.tenant_id, kb_id)


# ── 文件 ────────────────────────────────────────────────────────


@router.post(
    "/knowledge-bases/{kb_id}/documents",
    operation_id="documents_upload",
    status_code=status.HTTP_201_CREATED,
)
async def upload_document(
    kb_id: uuid.UUID,
    file: Annotated[UploadFile, File(description="單請求上傳上限 32MB，超過走分塊直傳")],
    principal: Annotated[Principal, Depends(RequireScope("knowledge:write"))],
) -> DocumentOut:
    """單請求 multipart 上傳（09 §3.1 的小檔路徑）。

    **讀檔在這裡、判定在 service**：讀取 request body 屬於「解析請求」（鐵則 3 允許），
    而「這份內容能不能收」是業務規則。

    ``file.content_type`` 刻意不使用——那是 client 自報的字串（見
    `services/knowledge/uploads.py`）。檔名同理只當顯示用，實際型別由內容決定。
    """
    content = await file.read()
    view = await run_orm(
        _documents.upload,
        principal.tenant_id,
        kb_id,
        filename=file.filename or "untitled",
        content=content,
        # 上傳者（2A-5）：ETL 失敗與 ready 的通知都寄給他。
        uploaded_by=principal.user_id,
    )
    return _document_out(view)


@router.get("/knowledge-bases/{kb_id}/documents", operation_id="documents_list")
async def list_documents(
    kb_id: uuid.UUID,
    principal: Annotated[Principal, Depends(RequireScope("knowledge:read"))],
) -> DocumentListOut:
    views = await run_orm(_documents.list_for_kb, principal.tenant_id, kb_id)
    return DocumentListOut(items=[_document_out(view) for view in views])


@router.get("/documents/{document_id}", operation_id="documents_get")
async def get_document(
    document_id: uuid.UUID,
    principal: Annotated[Principal, Depends(RequireScope("knowledge:read"))],
) -> DocumentOut:
    return _document_out(await run_orm(_documents.get, principal.tenant_id, document_id))


@router.post(
    "/documents/{document_id}/reingest",
    operation_id="documents_reingest",
    # 202：ETL 是非同步的（09 §2.3）。回 200 會讓 client 以為重跑已經完成，而它
    # 才剛進佇列——前端該做的是回去輪詢狀態，狀態碼就是那個提示。
    status_code=status.HTTP_202_ACCEPTED,
)
async def reingest_document(
    document_id: uuid.UUID,
    principal: Annotated[Principal, Depends(RequireScope("knowledge:write"))],
) -> DocumentOut:
    return _document_out(await run_orm(_documents.reingest, principal.tenant_id, document_id))


@router.delete(
    "/documents/{document_id}",
    operation_id="documents_delete",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_document(
    document_id: uuid.UUID,
    principal: Annotated[Principal, Depends(RequireScope("knowledge:write"))],
) -> None:
    await run_orm(_documents.delete, principal.tenant_id, document_id)
